hack/tools/metal.py

```diff
@@ -0,0 +1,171 @@
+      self.org_id = org_id
+      self.base = dirname(dirname(dirname(abspath(__file__))))
+      self.packet_manager = packet.Manager(auth_token=auth_token)
+      self.private_key_path = dirname(abspath(__file__))+"/private.key"
+      self.public_key_path = dirname(abspath(__file__))+"/public.key"
+      self.logger = self.set_logger(level)
+      self.ip = None
+
+   def set_logger(self, level):
+      logger = logging.getLogger('welder')
+      logger.setLevel(level)
+      c_handler = logging.StreamHandler()
+      c_handler.setLevel(level)
+      c_format = logging.Formatter('%(asctime)s %(levelname)s %(name)s - %(message)s', "%Y-%m-%d.%H:%M:%S")
+      c_handler.setFormatter(c_format)
+      logger.addHandler(c_handler)
+
+      return logger
+
+   def create_all(self, prj_name_or_id, dev_name, key_name, userdata=None):
+      project = self.create_new_project_if_not_exist(prj_name_or_id)
+      self.project = project
+      key = self.create_new_key(project.id, key_name)
+      self.key = key
+      if userdata == None:
+         userdata = self.create_user_data()
+      device = self.create_device(project.id, dev_name, key.id, userdata)
+      self.device = device
+      self.dev_id = device.id
+      self.wait_until_device_ready(device)
+
+      return self.ip
+
+   def create_new_project_if_not_exist(self, name_or_id):
+      project = None
+      try:
+         project = self.packet_manager.get_project(name_or_id)
+         self.logger.info(f"using found project {name_or_id}")
+         return project
+      except Exception:
+         pass
+
+      project = self.packet_manager.create_organization_project(
+          org_id=self.org_id,
+          name=name_or_id
+      )
+      self.logger.info(f"created project {name_or_id}")
+
+      return project
+
+   def create_new_key(self, project_id, name):
+      key = RSA.generate(2048)
+      with open(self.private_key_path, 'wb') as priv_file:
+          os.chmod(self.private_key_path, 0o600)
+          priv_file.write(key.exportKey('PEM'))
+
+      pubkey = key.publickey().exportKey('OpenSSH')
+      with open(self.public_key_path, 'wb') as pub_file:
+          pub_file.write(pubkey)
+
+      key = self.packet_manager.create_project_ssh_key(project_id, name, pubkey.decode("utf-8"))
+      self.logger.info(f"created key {key.label}")
+
+      return key
+
+   def create_user_data(self):
+      files = ["hack/scripts/bootstrap.sh", "hack/scripts/devpool.sh", "test/e2e/infra/userdata.sh"]
+      userdata = ""
+      for file in files:
+         with open(self.base+"/"+file) as f:
+             userdata += f.read()
+             userdata += "\n"
+
+      return userdata
+
+   def create_device(self, project_id, name, key_id, userdata):
+      device = self.packet_manager.create_device(project_id=project_id,
+                                     hostname=name,
+                                     plan='c1.small.x86', metro='sv',
```

Comment:
Would be good to make these configurable (with these values as defaults).

Reply:
yep, not as part of this tho 😄